feat: lazy load cms images /learn/ pages

fairlighteth opened the pull request:
# Summary

- Adds lazy loading for images coming from the CMS for the /learn/ and all /learn/* pages
- Adds lazy loading for <SVG by using a helper component <LazySVG
- Lighthouse optimizations (color contrast, image dimensions)
- Modified the `sitemap.xml` generating process to set the `lastMod` property for /learn/ articles to use the last modified date from the CMS. For the the /tokens/* pages it assumes daily updates. The other static pages for now will update also daily (on each build).
- Fixed /tokens and /tokens/* meta title/descriptions
- Fixed /learn/* (articles) meta descriptions: They use the `description` field from CMS, with a fallback in place.
- Adds Google Tag Manager by loading along the initAnalytics. It will **not** load by default, unless initialized as `initCowAnalyticsGoogle(true)`. Google Tag Manager may only load on cow.fi and **not** on swap.cow.fi 
- Consolidated the CategoryLinks component

elena-zh commented:
Hey @fairlighteth , when I open Learn page. I don't see that Topics images are loaded 
![image](https://github.com/user-attachments/assets/77e4d50e-0f17-421c-b153-b4992300a6f9)

They appear only when I press on any topic and then go back. Could you please check why this happens?

alfetopito commented on apps/cow-fi/hooks/useLazyLoadImages.tsx:
```diff
@@ -0,0 +1,69 @@
+export function useLazyLoadImages() {
+  const replaceImageUrls = React.useCallback((html: string): string => {
+    return html.replace(/<img([^>]*)src="([^"]*)"([^>]*)>/g, (match, before, src, after) => {
+      return `<img${before}data-src="${src}" src="data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='100%' height='100%' viewBox='0 0 1 1' preserveAspectRatio='none'%3E%3Crect width='1' height='1' fill='${LAZY_LOADING_CONFIG.placeholderColor.replace('#', '%23')}' /%3E%3C/svg%3E"${after}>`
```
This part is almost all duplicated with line `50`.
Can you extract this into a fn outside of the hook?

fairlighteth replied:
Addressed this now!

fairlighteth commented:
> Hey @fairlighteth , when I open Learn page. I don't see that Topics images are loaded ![image](https://private-user-images.githubusercontent.com/70885163/368593750-77e4d50e-0f17-421c-b153-b4992300a6f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3NzY4NDksIm5iZiI6MTcyNjc3NjU0OSwicGF0aCI6Ii83MDg4NTE2My8zNjg1OTM3NTAtNzdlNGQ1MGUtMGYxNy00MjFjLWIxNTMtYjQ5OTIzMDBhNmY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE5VDIwMDkwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE1MTRiNjZmN2Q0OTQ2ZDM4M2VlOWJjNDBkZTY1MDYxNTkzMTYxN2Y0ZTVmOGQwZWJiMzg1MmMwZWMzZTc3YmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7_OHCTinthjVIAEI4Vpsl-IiV2iPuGVutWzu_AmnKKs)
> 
> They appear only when I press on any topic and then go back. Could you please check why this happens?

Should be fixed 👍🏼

fairlighteth commented:
@elena-zh @alfetopito did updates to this PR so requesting a new review

shoom3301 commented on apps/cow-fi/pages/learn/articles/[[...page]].tsx:
```diff
@@ -75,8 +76,45 @@ export type ArticlesResponse = {
+  </CategoryLinks>
+)
+
+const renderArticles = (articles: any[]) => (
```
It supposed to be a component

shoom3301 commented on apps/cow-fi/pages/learn/articles/[[...page]].tsx:
```diff
@@ -75,8 +76,45 @@ export type ArticlesResponse = {
   }
 }
 
+const renderCategoryLinks = (allCategories: { name: string; slug: string }[]) => (
```
It supposed to be a component

shoom3301 commented on apps/cow-fi/components/LazySVG.tsx:
```diff
@@ -0,0 +1,67 @@
+
+  useEffect(() => {
+    if (!wrapperRef.current) {
+      console.log('wrapperRef is null')
```
Should we log it?

shoom3301 commented on apps/cow-fi/hooks/useLazyLoadImages.tsx:
```diff
@@ -0,0 +1,97 @@
+        }
+      }
+
+      initObserver(handleIntersect)
```
I think `initObserver` function doesn't work right.
So, it creates `IntersectionObserver` only once regardless of a context.
It means that if you call the `useLazyLoadImages()` hook once, in the next time it won't work and `handleIntersect` won't be called

shoom3301 commented on apps/cow-fi/hooks/useLazyLoadImages.tsx:
```diff
@@ -0,0 +1,97 @@
+  )
+}
+
+export function useLazyLoadImages() {
```
Can we just use [native lazy-loading](https://developer.mozilla.org/en-US/docs/Web/Performance/Lazy_loading) instead?

shoom3301 replied:
Or https://github.com/Aljullu/react-lazy-load-image-component

fairlighteth replied:
The issue was that content was being loaded from markdown/cms and  we need to deal what that content

shoom3301 commented on apps/cow-fi/hooks/useLazyLoadImages.tsx:
```diff
@@ -0,0 +1,97 @@
+}
+
+export function useLazyLoadImages() {
+  const placeholderSrc = useRef(getPlaceholderSrc(LAZY_LOADING_CONFIG.placeholderColor)).current
```
`placeholderSrc` is actually static, we don't need `useRef` here

shoom3301 commented on apps/cow-fi/pages/learn/[article].tsx:
```diff
@@ -119,6 +116,29 @@ function getRandomArticles(articles: Article[], count: number): Article[] {
+  const { replaceImageUrls, LazyImage } = useLazyLoadImages()
+
+  // Preprocess the HTML content
+  const processedContent = React.useMemo(() => {
```
Just to be consistent with the rest of code-base, please use it as `import { useMemo } from 'react'`

elena-zh reviewed (commented):
Hey @fairlighteth , everything is great!
However, there is one new thing appeared with the main menu in a mobile view:
 - white periods are displayed from the right side of every line
 - and the menu became scrollable (a bit) while it has fixed position on Prod
 
![image](https://github.com/user-attachments/assets/1b3d595a-1103-4b21-aa58-13090cc13f26)
![image](https://github.com/user-attachments/assets/350157fc-4e3c-42ed-8895-b4a8452473fc)

fairlighteth commented:
@shoom3301 I should have addressed your comments.
@elena-zh the dots should be gone, could you double check (I could not re-produce myself). As for scrolling I couldn't reproduce either.

Did some refactoring for the CategoryLinks

elena-zh commented:
The build has failed :(

fairlighteth commented:
@elena-zh pushed a fix for that, please review
